tests/lib/fakeinstaller/main.go

```diff
@@ -0,0 +1,282 @@
+func waitChange(chgId string) error {
+	cli := client.New(nil)
+	for {
+		chg, err := cli.Change(chgId)
+		if err != nil {
+			return err
+		}
+
+		if chg.Err != "" {
+			return errors.New(chg.Err)
+		}
+		if chg.Ready {
+			return nil
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// TODO laidoutStructs is used to get the devices, when encryption is
+// happening maybe we need to find the information differently.
+func postSystemsInstallFinish(cli *client.Client,
+	details *client.SystemDetails, bootDevice string,
+	laidoutStructs []gadget.OnDiskStructure) error {
+
+	vols := make(map[string]*gadget.Volume)
+	for volName, gadgetVol := range details.Volumes {
+		laidIdx := 0
+		for i := range gadgetVol.Structure {
+			// TODO mbr is special, what is the device for that?
+			var device string
+			if gadgetVol.Structure[i].Role == "mbr" {
+				device = bootDevice
+			} else {
+				device = laidoutStructs[laidIdx].Node
+				laidIdx++
+			}
+			gadgetVol.Structure[i].Device = device
+		}
+		vols[volName] = gadgetVol
+	}
+
+	// Finish steps does the writing of assets
+	opts := &client.InstallSystemOptions{
+		Step:      client.InstallStepFinish,
+		OnVolumes: vols,
+	}
+	chgId, err := cli.InstallSystem(details.Label, opts)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Change %s created\n", chgId)
+	return waitChange(chgId)
+}
+
+// createAndMountFilesystems creates and mounts filesystems. It returns
+// an slice with the paths where the filesystems have been mounted to.
+func createAndMountFilesystems(bootDevice string, volumes map[string]*gadget.Volume) ([]string, error) {
+	// only support a single volume for now
+	if len(volumes) != 1 {
+		return nil, fmt.Errorf("got unexpected number of volumes %v", len(volumes))
+	}
+
+	disk, err := disks.DiskFromDeviceName(bootDevice)
+	if err != nil {
+		return nil, err
+	}
+	vol := firstVol(volumes)
+
+	var mountPoints []string
+	for _, stru := range vol.Structure {
+		if stru.Label == "" || stru.Filesystem == "" {
+			continue
+		}
+
+		part, err := disk.FindMatchingPartitionWithPartLabel(stru.Label)
+		if err != nil {
+			return nil, err
+		}
+		// XXX: reuse
+		// gadget/install/content.go:mountFilesystem() instead
+		// (it will also call udevadm)
+		if err := mkfs.Make(stru.Filesystem, part.KernelDeviceNode, stru.Label, 0, 0); err != nil {
+			return nil, err
+		}
+
+		// mount
+		mountPoint := runMntFor(stru.Label)
+		if err := os.MkdirAll(mountPoint, 0755); err != nil {
+			return nil, err
+		}
+		// XXX: is there a better way?
+		if output, err := exec.Command("mount", part.KernelDeviceNode, mountPoint).CombinedOutput(); err != nil {
+			return nil, osutil.OutputErr(output, err)
+		}
+		mountPoints = append(mountPoints, mountPoint)
+	}
+
+	return mountPoints, nil
+}
+
+func unmountFilesystems(mntPts []string) error {
+	for _, mntPt := range mntPts {
+		if output, err := exec.Command("umount", mntPt).CombinedOutput(); err != nil {
+			return osutil.OutputErr(output, err)
```

Comment:
(nitpick) this should probably best effort, i.e. try to umount as much as possible and return the first error (or conact of al error, this is tricky and annoying) - but fine for now.